keps/sig-cli/3104-introduce-kuberc/README.md

```diff
@@ -0,0 +1,849 @@
+* Enable users to define kubectl command aliases.
+* Enable users to define kubectl default flags.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We propose introducing a new file that will be versioned and intended for user provided preferences. This new file is entirely opt-in.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+| Risk | Impact | Mitigation |
+| --- | --- | --- |
+| Confusing users with a new config file | Low | Documentation and education |
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The file will default to being located in `~/.kube/kuberc`. A flag will allow overriding this default location with a path i.e. `kubectl --kuberc /var/kube/rc`.
+
+Three initial top level keys are proposed.
+
+* `apiVersion` to determine the version of the config.
+* `commandAliases` for users to declare their own aliases for commands with flags and values.
+* `commandOverrides` for users to set default flags to apply to commands.
+
+`commandAliases` will take precedence when looking through the code execution path i.e. aliases -> builtin -> plugins. This will be expanded early in the kubectl invocation and used to replace argv. Additional flags and values will be appended to the end of the aliased command.
```

Comment:
aliases override builtins? that's surprising to me

Reply:
That's fair. My initial thought was if you wanted to set defaults for a command you could create an alias for it but that is really covered by the overrides.

I'll change that.

Reply:
I agree with Jordan on this one, we need to ensure that builtin commands stay builtin commands, otherwise we'll introduce a new set of problems where users will start reporting problems with builtin commands that are shadowed with aliases. Having aliases after builtins will also be identical how we work with plugins, which also can't shadow builtins.